chat-android/src/main/java/com/ably/chat/RoomLifecycleManager.kt

```diff
@@ -0,0 +1,477 @@
+package com.ably.chat
+
+import io.ably.lib.realtime.ChannelState
+import io.ably.lib.types.AblyException
+import io.ably.lib.types.ErrorInfo
+import io.ably.lib.util.Log.LogHandler
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.suspendCancellableCoroutine
+import io.ably.lib.realtime.Channel as AblyRealtimeChannel
+
+/**
+ * An interface for features that contribute to the room status.
+ */
+interface ContributesToRoomLifecycle : EmitsDiscontinuities, HandlesDiscontinuity {
+    /**
+     * Gets the channel on which the feature operates. This promise is never
+     * rejected except in the case where room initialization is canceled.
+     */
+    override val channel: AblyRealtimeChannel
+
+    /**
+     * Gets the ErrorInfo code that should be used when the feature fails to attach.
+     * @returns The error that should be used when the feature fails to attach.
+     */
+    val attachmentErrorCode: ErrorCodes
```

Comment:
`attachmentErrorCode` is a strongly typed errorCode value for each contributor/room feature.
This is one of the reason why error codes are represented as a part of `ErrorCodes` enum.
If it's a simple `Int` value, there's a possibility of assigning wrong error code value. 
I think we should encourage compile-time type checking wherever we can.